pkg/diff/sql_graph.go

```diff
@@ -4,64 +4,124 @@ import (
+	diffType diffType
+}
+
+func buildSchemaObjVertexId(objType string, id string, diffType diffType) sqlVertexId {
+	return schemaObjSqlVertexId{
+		objType:  objType,
+		objId:    id,
+		diffType: diffType,
+	}
+}
+
+func (s schemaObjSqlVertexId) String() string {
+	return fmt.Sprintf("%s:%s:%s", s.objType, s.objId, s.diffType)
+}
+
 type sqlVertex struct {
-	ObjId      string
-	Statements []Statement
-	DiffType   diffType
+	// id is used to identify the sql vertex
+	id sqlVertexId
+
+	// priority is used to determine if the sql vertex should be included sooner or later in the topological
+	// sort of the graph
+	priority sqlPriority
+
+	// statements is the set of statements to run for this vertex
+	statements []Statement
 }
 
 func (s sqlVertex) GetId() string {
-	return fmt.Sprintf("%s_%s", s.DiffType, s.ObjId)
+	return s.id.String()
 }
 
-func buildTableVertexId(name schema.SchemaQualifiedName) string {
-	return fmt.Sprintf("table_%s", name)
+func (s sqlVertex) GetPriority() int {
+	// Prioritize adds/alters over deletes. Weight by number of statements. A 0 statement delete should be
+	// prioritized over a 1 statement delete
+	return len(s.statements) * int(s.priority)
 }
 
-func buildIndexVertexId(name schema.SchemaQualifiedName) string {
-	return fmt.Sprintf("index_%s", name)
+// dependency indicates an edge between the SQL to resolve a diff for a source schema object and the SQL to resolve
+// the diff of a target schema object
+//
+// Most SchemaObjects will have two nodes in the SQL graph: a node for delete SQL and a node for add/alter SQL.
+// These nodes will almost always be present in the sqlGraph even if the schema object is not being deleted (or added/altered).
+// If a node is present for a schema object where the "diffType" is NOT occurring, it will just be a no-op (no SQl statements)
+type dependency struct {
+	source sqlVertexId
+	target sqlVertexId
 }
 
-// sqlGraph represents two dependency webs of SQL statements
-type sqlGraph graph.Graph[sqlVertex]
+type dependencyBuilder struct {
+	base sqlVertexId
+}
+
+func mustRun(id sqlVertexId) dependencyBuilder {
+	return dependencyBuilder{
+		base: id,
+	}
+}
 
-// union unions the two AddsAndAlters graphs and, separately, unions the two delete graphs
-func (s *sqlGraph) union(sqlGraph *sqlGraph) error {
-	if err := (*graph.Graph[sqlVertex])(s).Union((*graph.Graph[sqlVertex])(sqlGraph), mergeSQLVertices); err != nil {
-		return fmt.Errorf("unioning the graphs: %w", err)
+func (d dependencyBuilder) before(id sqlVertexId) dependency {
+	return dependency{
+		source: d.base,
+		target: id,
 	}
-	return nil
 }
 
-func mergeSQLVertices(old, new sqlVertex) sqlVertex {
-	return sqlVertex{
-		ObjId:      old.ObjId,
-		DiffType:   old.DiffType,
-		Statements: append(old.Statements, new.Statements...),
+func (d dependencyBuilder) after(id sqlVertexId) dependency {
+	return dependency{
+		source: id,
+		target: d.base,
+	}
+}
+
+// sqlGraph represents two dependency webs of SQL statements
```

Comment:
nit: "two" graphs? Where does that come from?

Reply:
oops, that long gone